app/master/cluster_master.py

```diff
@@ -62,12 +65,24 @@ def api_representation(self):
             'slaves': slaves_representation,
         }
 
-    def builds(self):
+    def builds(self, offset: int=None, limit: int=None) -> List['Build']:
         """
-        Returns a list of all builds
-        :rtype: list[Build]
+        Returns a list of all builds. Both offset and limit are assumed to be nonnegative at this point
+        since the corrections or negative offsets and limits are done when the query parameters are first initialized.
+        :param offset: The starting index of the requested build
+        :param limit: The number of builds requested
         """
-        return self._all_builds_by_id.values()
+        num_builds = len(self._all_builds_by_id)
+        offset = offset or 0
+        # Don't return more than the MAX_LIMIT even if we have more builds than that
+        limit = limit if limit is not None else min(num_builds, pagination_constants['MAX_LIMIT'])
+
+        # Reset the offset to 0 if it is out of range
+        starting_index = offset if offset < num_builds else 0
+        ending_index = min((starting_index + limit), num_builds)
+
+        requested_builds = islice(self._all_builds_by_id, starting_index, ending_index)
```

Comment:
Writing this here for visibility. I don't think anyone should actually fix this right now, I just want to make sure this "bug" is known.

I noticed that 
1. `_all_builds_by_id` is an OrderedDict, not an array
2. `islice` returns builds things from an OrderedDict in the order they were added to the dict, not in the numerical order of the keys

This means if somehow builds got added to the OrderedDict out of numerical order, `islice` would return the builds out of order.

For example, say we got 3 requests all at the same time. We created build ids 0,1,2 for them but we added them to the OrderedDict in the order 2,1,0. Then if I made a request to clusterrunner with `offset=0,limit=1` I would get build 2 instead of build 0.

That being said, this behavior is actually impossible right now. The "bug" I described relies on builds being added to the dict in a different order than they are created, but the current code is setup in a way that that can never happen. Builds are created and then immediately added to the dict in `handle_request_for_new_build`. This happens on tornado's single thread, so there's no possibility of threads interleaving operations in a way that causes this bug.

So ... nothing to do right now. Just wanted to write this down. At some point someone should maybe change `_all_builds_by_id` to be an array though, which would fix this potential bug.